### PR TITLE
Add a new mockpatch fixture

### PR DESCRIPTION
**Pull request by jd**
This fixture provides an easy usage for mock (unittest.mock in
Python 3).

It has been imported from oslotest in order to consolidate our fixtures
upstream.

### CODE REVIEW

**Comment by rbtcollins** [comment]
So this is adding new contributors based on the header in the file you've added. Thats fine but please update AUTHORS and see COPYING to make sure you are able to contribute this code to fixtures. In particular fixtures is dual-license BSD and Apache2, so we can't accept Apache2 only code.

File: fixtures/_fixtures/mockpatch.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        self.obj = obj
+        self.attr = attr
+        self.kwargs = kwargs
+        self.new = new
+
+    def setUp(self):
+        super(MockPatchObject, self).setUp()
+        _p = mock.patch.object(self.obj, self.attr, self.new, **self.kwargs)
+        self.mock = _p.start()
+        self.addCleanup(_p.stop)
```
Comment (rbtcollins) [inline fixtures/_fixtures/mockpatch.py]: There's a small race here between p.start() and addCleanup. You should call addCleanup first:
self.addCleanup(_p.stop)
self.mock = _p.start()

If p.start() errors badly stop will be called spuriously which may result in spurious errors, but this is better than potentially leaking a globally applied mock.

File: fixtures/_fixtures/mockpatch.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self, obj, new=mock.DEFAULT, **kwargs):
+        self.obj = obj
+        self.kwargs = kwargs
+        self.new = new
+
+    def setUp(self):
+        super(MockPatch, self).setUp()
+        _p = mock.patch(self.obj, self.new, **self.kwargs)
+        self.mock = _p.start()
+        self.addCleanup(_p.stop)
```
Comment (rbtcollins) [inline fixtures/_fixtures/mockpatch.py]: likewise.

File: fixtures/_fixtures/mockpatch.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        """
+        self.obj = obj
+        self.kwargs = kwargs
+
+    def setUp(self):
+        super(MockPatchMultiple, self).setUp()
+        _p = mock.patch.multiple(self.obj, **self.kwargs)
+        self.mock = _p.start()
+        self.addCleanup(_p.stop)
```
Comment (rbtcollins) [inline fixtures/_fixtures/mockpatch.py]: And here. Further, the differences between these three classes appear to be in two things: the method called on mock, and the marshalling of variables in the __init__. I think a little bit of refactoring would let you have a single setUp method shared by them all reducing code size and duplication. E.g. one way would be to say

self._get_p = partial(mock.patch.multiple(obj, **kwargs)

in __init__, and then setUp just calls self._get_p()

Comment (rbtcollins) [inline fixtures/_fixtures/mockpatch.py]: That would also let the test surface area be shrunk - we could focus on checking the implementation of _get_p in each class.

File: fixtures/tests/_fixtures/test_mockpatch.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+class Foo(object):
+    def bar(self):
+        return self
+
+
+def mocking_bar(self):
+    return 'mocked!'
+
+
+class TestMockPatch(testtools.TestCase, TestWithFixtures):
```
Comment (rbtcollins) [inline fixtures/tests/_fixtures/test_mockpatch.py]: TestWithFixtures isn't needed when using testtools.TestCase.

File: requirements-py3.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+testtools>=0.9.22
```
Comment (rbtcollins) [inline requirements-py3.txt]: This file isn't needed. Mock is installable on python3 with no harmful side effects - its simpler to just have the one requirements file.

**Comment by rbtcollins** [comment]
Sorry about the mass delay reviewing this - hit my mailbox mid conference and just got back to the backlog now.

**Comment by jd** [comment]
Hi @rbtcollins !

I've merge your suggestions upstream and now I'm back with that PR. I don't know what to do about the licensing issue, so hint appreciated.

**Comment by rbtcollins** [comment]
I think if you look at the git history you can identify the contributors and verify that they're happy to have their code dual licensed under BSD and ApacheV2 that that will be sufficient. I hate to have to ask, but thats licensing :/

**Comment by jd** [comment]
You should have received a copy of all agreements from all developers. Do you need anything more?

**Comment by rbtcollins** [comment]
Looking good. The new dep is a little odd; I'm not sure whether to a) make it conditional, which still has very spotty ecosystem support or b) make it optional where folk that want to use this fixture with mock need to install it themselves/separately. I think I'm going to remove that dep and tweak the code to fail gracefully.

**Comment by rbtcollins** [comment]
Merged with tweaks:
 - added documentation
 - made the mock feature not depend on mock, but added it to test-requirements.

**Pull request closed without merging**